diktat-rules/src/main/kotlin/org/cqfn/diktat/ruleset/rules/files/FileStructureRule.kt

```diff
@@ -107,56 +125,127 @@ class FileStructureRule(private val configRules: List<RulesConfig>) : Rule("file
+
+        if (sortedImportsGroups.flatten() != imports) {
             FILE_UNORDERED_IMPORTS.warnAndFix(configRules, emitWarn, isFixMode, fileName, node.startOffset, node) {
-                rearrangeImports(node, imports, sortedImports)
+                rearrangeImports(node, imports, sortedImportsGroups)
             }
         }
     }
 
-    private fun rearrangeImports(node: ASTNode, imports: List<ASTNode>, sortedImports: List<ASTNode>) {
+    private fun rearrangeImports(node: ASTNode, imports: List<ASTNode>, sortedImportsGroups: List<List<ASTNode>>) {
         require(node.elementType == IMPORT_LIST)
         // move all commented lines among import before imports block
         node.getChildren(TokenSet.create(EOL_COMMENT))
-                .forEach {
-                    node.treeParent.addChild(it.clone() as ASTNode, node)
-                    node.treeParent.addChild(PsiWhiteSpaceImpl("\n"), node)
-                }
+            .forEach {
+                node.treeParent.addChild(it.clone() as ASTNode, node)
+                node.treeParent.addChild(PsiWhiteSpaceImpl("\n"), node)
+            }
 
         node.removeRange(imports.first(), imports.last())
-        sortedImports.forEachIndexed { index, importNode ->
-            if (index != 0) {
-                node.addChild(PsiWhiteSpaceImpl("\n"), null)
+        sortedImportsGroups.filterNot { it.isEmpty() }
+            .run {
+                forEachIndexed { groupIndex, group ->
+                    group.forEachIndexed { index, importNode ->
+                        node.addChild(importNode, null)
+                        if (index != group.size - 1) {
+                            node.addChild(PsiWhiteSpaceImpl("\n"), null)
+                        }
+                    }
+                    if (groupIndex != size - 1) {
+                        node.addChild(PsiWhiteSpaceImpl("\n\n"), null)
+                    }
+                }
             }
-            node.addChild(importNode, null)
-        }
     }
 
-    private fun ASTNode.getSiblingBlocks(copyrightComment: ASTNode?,
-                                         headerKdoc: ASTNode?,
-                                         fileAnnotations: ASTNode?,
-                                         packageDirectiveNode: ASTNode
+    private fun ASTNode.getSiblingBlocks(
+        copyrightComment: ASTNode?,
+        headerKdoc: ASTNode?,
+        fileAnnotations: ASTNode?,
+        packageDirectiveNode: ASTNode
     ): Pair<ASTNode?, ASTNode> = when (elementType) {
         BLOCK_COMMENT -> null to listOfNotNull(headerKdoc, fileAnnotations, packageDirectiveNode).first()
         KDOC -> copyrightComment to (fileAnnotations ?: packageDirectiveNode)
         FILE_ANNOTATION_LIST -> (headerKdoc ?: copyrightComment) to packageDirectiveNode
         else -> error("Only BLOCK_COMMENT, KDOC and FILE_ANNOTATION_LIST are valid inputs.")
     }
 
+    private data class ImportGroups(
```

Comment:
you are not using anything from a data class here 
simply generating extra code

Reply:
Yes, I must've thought that eventually there will be more grouping-specific logic. I'll remove data class.